akka-persistence/src/main/scala/akka/persistence/AtLeastOnceDelivery.scala

```diff
@@ -380,6 +384,8 @@ abstract class AbstractPersistentActorWithAtLeastOnceDelivery extends AbstractPe
    * This method will throw [[AtLeastOnceDelivery.MaxUnconfirmedMessagesExceededException]]
    * if [[numberOfUnconfirmed]] is greater than or equal to [[#maxUnconfirmedMessages]].
    */
-  def deliver(destination: ActorPath, deliveryIdToMessage: akka.japi.Function[java.lang.Long, Object]): Unit =
+  def deliver(destination: ActorPath, deliveryIdToMessage: akka.japi.function.Function[java.lang.Long, Object]): Unit =
     super.deliver(destination, id ⇒ deliveryIdToMessage.apply(id))
+
+  // FIXME @deprecated deliver with akka.japi.Function. How would java8 know which one to pick when param is lambda?
```

Comment:
good point, um I don't know what to do.

What would happen if we introduced the variance in the old functions?
We had some thought about that it would be binary compatible but not source compatible if used from Scala. Would that be acceptable for 2.4? Then we would have to break binary compatibility of streams when moving it to 2.4, but that might be alright.